Data/ByteString/Short/Internal.hs

```diff
@@ -446,6 +554,781 @@ concat sbss =
+last = \sbs -> case null sbs of
+  True -> error "empty ShortByteString"
+  False -> indexWord8Array (asBA sbs) (length sbs - 1)
+{-# INLINE last #-}
+
+-- | /O(n)/ Extract the elements after the head of a ShortByteString, which must be non-empty.
+-- An exception will be thrown in the case of an empty ShortByteString.
+--
+-- Note: copies the entire byte array
+--
+-- @since 0.11.3.0
+tail :: ShortByteString -> ShortByteString
+tail = \sbs -> 
+  let l = length sbs
+      nl = l - 1
+  in case null sbs of
+      True -> error "empty ShortByteString"
+      False -> create nl $ \mba -> copyByteArray (asBA sbs) 1 mba 0 nl
+{-# INLINE tail #-}
+
+-- | /O(1)/ Extract the first element of a ShortByteString, which must be non-empty.
+-- An exception will be thrown in the case of an empty ShortByteString.
+--
+-- @since 0.11.3.0
+head :: ShortByteString -> Word8
+head = \sbs -> case null sbs of
+  True -> error "empty ShortByteString"
+  False -> indexWord8Array (asBA sbs) 0
+{-# INLINE head #-}
+
+-- | /O(n)/ Return all the elements of a 'ShortByteString' except the last one.
+-- An exception will be thrown in the case of an empty ShortByteString.
+--
+-- Note: copies the entire byte array
+--
+-- @since 0.11.3.0
+init :: ShortByteString -> ShortByteString
+init = \sbs ->
+  let l = length sbs
+      nl = l - 1
+  in case null sbs of
+      True -> error "empty ShortByteString"
+      False -> create nl $ \mba -> copyByteArray (asBA sbs) 0 mba 0 nl
+{-# INLINE init #-}
+
+
+-- ---------------------------------------------------------------------
+-- Transformations
+
+-- | /O(n)/ 'map' @f xs@ is the ShortByteString obtained by applying @f@ to each
+-- element of @xs@.
+--
+-- @since 0.11.3.0
+map :: (Word8 -> Word8) -> ShortByteString -> ShortByteString
+map f = \sbs ->
+    let l = length sbs
+        ba = asBA sbs
+    in create l (\mba -> go ba mba 0 l)
+  where
+    go :: BA -> MBA s -> Int -> Int -> ST s ()
+    go !ba !mba !i !l
+      | i >= l = return ()
+      | otherwise = do
+          let w = indexWord8Array ba i
+          writeWord8Array mba i (f w)
+          go ba mba (i+1) l
+
+
+-- | /O(n)/ 'reverse' @xs@ efficiently returns the elements of @xs@ in reverse order.
+--
+-- @since 0.11.3.0
+reverse :: ShortByteString -> ShortByteString
+reverse = \sbs ->
+    let l = length sbs
+        ba = asBA sbs
+    in create l (\mba -> go ba mba 0 l)
+  where
+    go :: BA -> MBA s -> Int -> Int -> ST s ()
+    go !ba !mba !i !l
+      | i >= l = return ()
+      | otherwise = do
+          let w = indexWord8Array ba i
+          writeWord8Array mba (l - 1 - i) w
+          go ba mba (i+1) l
+
+
+-- | /O(n)/ The 'intercalate' function takes a 'ShortByteString' and a list of
+-- 'ShortByteString's and concatenates the list after interspersing the first
+-- argument between each element of the list.
+--
+-- @since 0.11.3.0
+intercalate :: ShortByteString -> [ShortByteString] -> ShortByteString
+intercalate s = concat . List.intersperse s
```

Comment:
It's faster to implement directly, cf. 
https://github.com/haskell/bytestring/blob/6ff6ed4b580ec7a919ac7abbd9fbb64a53353722/Data/ByteString.hs#L1209-L1227

Reply:
Are you sure? Did you benchmark it?